### PR TITLE
[INFINITY-2379] Extend state store API to get/set files

### PR DESCRIPTION
**Pull request by NimaVaziri**
This adds to the `StateStore` endpoints the ability to upload files, download them, and list the files the `StateStore` holds at any given time.

- [x] add tests

### CODE REVIEW

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/api/StateResource.java [hunk 1]
```diff
@@ -94,6 +97,70 @@ public Response getPropertyKeys() {
...
+        String fileName = fileDetails.getFileName();
+        if (uploadedInputStream == null || fileName == null) {
+            return ResponseUtils.plainResponse(
+                    "Specify 'file' as a parameter in your request",
+                    Response.Status.BAD_REQUEST
+            );
+        }
+
+        try {
+            if (stateStore.hasProperty(StateStoreUtils.FILE_NAME_PREFIX + fileName)) {
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/api/StateResource.java]: Should the prior content be overwritten in this case? If so, then may as well omit the new StateStore.hasProperty() call as this is the only place using it at the moment.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/state/StateStore.java [hunk 2]
```diff
@@ -452,6 +452,22 @@ public void clearAllData() throws StateStoreException {
...
+     */
+    public boolean hasProperty(String key) throws StateStoreException {
+        try {
+            final String path = PersisterUtils.join(PROPERTIES_PATH_NAME, key);
+            logger.debug("Determining whether key: {} exists at path: {}", key, path);
+            // persister throws an exception if the path (i.e. key) doesn't exist.
+            persister.get(path);
+            return true;
+        } catch (PersisterException e) {
+            logger.info("Property {} doesn't exist in the state store", key);
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/state/StateStore.java]: Should check for e.getReason() == Reason.NOT_FOUND. see example at L446

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/api/StateResource.java [hunk 3]
```diff
@@ -94,6 +97,70 @@ public Response getPropertyKeys() {
...
+                String msg = String.format("File %s is already in the state store", fileName);
+                logger.info(msg);
+                return ResponseUtils.plainOkResponse(msg);
+            }
+            logger.info("Storing {}", fileName);
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/api/StateResource.java]: Should have a sanity check against the metadata that the advertised size isn't too big (maybe a field in fileDetails has this?)

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/state/StateStoreUtils.java [hunk 4]
```diff
@@ -304,4 +310,45 @@ public static void setLastCompletedUpdateType(
             return ConfigurationUpdater.UpdateResult.DeploymentType.valueOf(value);
         }
     }
+
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/state/StateStoreUtils.java]: Looks like these methods are only used in StateResource. Can just move them to private methods in StateResource.

File: sdk/scheduler/src/test/java/com/mesosphere/sdk/state/StateStoreUtilsTest.java [hunk 5]
```diff
@@ -32,6 +33,8 @@
  */
 public class StateStoreUtilsTest {
```
Comment (nickbp) [inline sdk/scheduler/src/test/java/com/mesosphere/sdk/state/StateStoreUtilsTest.java]: Missing unit tests against the new StateResource calls?

Should also exercise the malicious large content cases described above:
- `Content-Length` too big: fail fast.
- `Content-Length` small but actual content too big: fail during streaming.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/state/StateStoreUtils.java [hunk 6]
```diff
@@ -304,4 +310,45 @@ public static void setLastCompletedUpdateType(
...
+    /**
+     * Stores the file in the state store.
+     * @param stateStore The state store to store the file in.
+     * @param fileName The name of the file to store.
+     * @param uploadedInputStream The input stream holding the content of the file.
+     */
+    public static void storeFile(StateStore stateStore, String fileName, InputStream uploadedInputStream)
+            throws StateStoreException, IOException {
+        StringWriter stringWriter = new StringWriter();
+        IOUtils.copy(uploadedInputStream, stringWriter, FILE_ENCODING);
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/state/StateStoreUtils.java]: Need to limit the size of this copy. This check is needed in addition to any size check against the metadata above.
 Malicious input could have a `Content-Length` of 1024byte and an actual content length of infinity. E.g. piping /dev/zero into HTTP.

**Comment by NimaVaziri** [comment]
These commits were separated into their own PR but originally it was for a journal node case with HDFS where the journal nodes have to upload a file as part of handling a replace operation (which is coming up in another PR).

File: sdk/scheduler/src/test/java/com/mesosphere/sdk/state/StateStoreUtilsTest.java (hunk 5 above)
Comment (nickbp) [reply sdk/scheduler/src/test/java/com/mesosphere/sdk/state/StateStoreUtilsTest.java]: (These cases could probably be done by mocking InputStream and/or FormDataContentDisposition)

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/api/StateResource.java (hunk 1 above)
Comment (NimaVaziri) [reply sdk/scheduler/src/main/java/com/mesosphere/sdk/api/StateResource.java]: Yeah, this restricts the more general use case...will omit.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/state/StateStoreUtils.java (hunk 6 above)
Comment (NimaVaziri) [reply sdk/scheduler/src/main/java/com/mesosphere/sdk/state/StateStoreUtils.java]: Ah, in memory of heartbleed.

**Comment by NimaVaziri** [comment]
@nickbp addressed your comments, any other ones?

**Comment by nickbp** [comment]
Nope!

**Pull request merged**